File: twoPointers/validPalindrome.py
#my crack at valid palindrome
class Solution(object):
    def lowerCaseIt(self, char):
        if char >= "A" and char <= "Z":
            return chr(ord(char) + 32)
        return char

    def isAlphaNum(self, char):
        if char >= "A" and char <= "Z":
            return True
        elif char >= "a" and char <= "z":
            return True
        elif char >= "0" and char <= "9":
            return True
        return False
    
    def isPalindrome(self, s):
        left = 0
        right = len(s) - 1
        while left <= right:
            if self.isAlphaNum(s[left]) and self.isAlphaNum(s[right]):
                if self.lowerCaseIt(s[left]) != self.lowerCaseIt(s[right]):
                    return False
                left += 1
                right -= 1
            #if not, adjust left and right indexes as needed for next cycle
            elif not self.isAlphaNum(s[left]) and self.isAlphaNum(s[right]):
                left += 1
            elif self.isAlphaNum(s[left]) and not self.isAlphaNum(s[right]):
                right -= 1
            else:
                left += 1
                right -= 1
        return True

File: twoPointers/validPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        //use two pointers(indexes)
        //start on both sides of the string
        //adjust both pointers until both are pointing
        //to alphanum chars
        //if different, then we know this isn't a palindrome
        //continue this entire process until the two pointers
        //cross over
        int l = 0;
        int r = s.size() - 1;

        while (l <= r){
            while (l <= r && isalnum(s[l]) == 0)
                l++;
            while (l <= r && isalnum(s[r]) == 0)
                r--;
            if(l <= r && tolower(s[l]) != tolower(s[r]))
                return false;
            l++;
            r--;
        }
        return true;
    }
};
